### PR TITLE
add content filter to filter the article by keyword expression

### PR DESCRIPTION
**Pull request by Jinwei1987**
add content filter to filter the article by keyword expression
it will be able to support expression like 'key1 | key2 | key3' or more complex one like '(key1 & key2) | (key3 & key4)'

### CODE REVIEW

File: newsplease/pipeline/pipelines.py [hunk 1]
```diff
@@ -570,6 +571,147 @@ def process_item(self, item, spider):
                 return item
 
 
+class ContentFilter(object):
+
+    class BiOp(object):
```
Comment (fhamborg) [inline newsplease/pipeline/pipelines.py]: Could you please use a better understandable class name?

Comment (fhamborg) [reply newsplease/pipeline/pipelines.py]: and briefly state what this class is doing

File: newsplease/config/config.cfg [hunk 2]
```diff
@@ -263,6 +263,13 @@ end_date = '2999-12-31 00:00:00'
 strict_mode = False
 
 
+[ContentFilter]
+
+filter_predicates = {'title': 'cloud computing',
+                     'text': 'google | amazon'}
```
Comment (fhamborg) [inline newsplease/config/config.cfg]: Can you add a brief comment to explain how this is to be used? E.g., that you can use the term filter on either the title or the text field.

File: newsplease/config/config.cfg [hunk 3]
```diff
@@ -263,6 +263,13 @@ end_date = '2999-12-31 00:00:00'
...
+pass_condition = title or text
```
Comment (fhamborg) [inline newsplease/config/config.cfg]: Please specify as a comment any other values that one can put here besides title or text.

File: newsplease/config/config.cfg (hunk 2 above)
Comment (fhamborg) [reply newsplease/config/config.cfg]: Also, can one use pure strings or also regex?

**Comment by fhamborg** [comment]
Hi! Thank you for the PR! I've added a few comments in order to clear up a few minor things. I really appreciate your PR and support :-)

File: newsplease/config/config.cfg (hunk 2 above)
Comment (Jinwei1987) [reply newsplease/config/config.cfg]: Pure string without &, | and ( is supported. 
regex is not supported. I feel logic expression is easier to use comparing with regex. with nested logic expression, there will be rare case where regex is needed

**Comment by Jinwei1987** [comment]
> Hi! Thank you for the PR! I've added a few comments in order to clear up a few minor things. I really appreciate your PR and support :-)

I have added the description of the class and moments to explain the use cases

File: newsplease/pipeline/pipelines.py (hunk 1 above)
Comment (Jinwei1987) [reply newsplease/pipeline/pipelines.py]: have added the comments

**Comment by fhamborg** [comment]
We finally had a news-please core developer meeting so that I can give an update on this. While we still want to add the filter functionality, we think that using regex solely is the most suitable approach. Compared to the approach in this PR, regex not only supports basic logic functionality but much more in terms of string and pattern matching. Thus, we cannot merge the PR in this form into the master branch. Still, thank you for the PR! If you are considering removing the biop functionality and use regex instead, please just let me know. Thank you!

**Pull request closed without merging**